notes.md

```diff
@@ -0,0 +1,179 @@
+- We want to receive websocket frames.
+  - We want to backpressure the sender. For context, say our service proxies frames to other tcp servers. Furthermore,
+    one of those servers is slow. We want the client to slow down their rate without an explicit higher layer protocol
+    message. However, we can optimize this so that a single slow downstream does backpressure all other frames going to
+    other destinations. We maintain a queue of frames and only backpressure when queue is full. (Without an explicit
+    higher layer flow control or protocol other TCP, this will only mitigate the problem. Frames for different
+    downstreams are multiplexed over a single TCP connection, so once queue is full because a single slow consumer than 
+    we block all other frame processing, a.k.a Head-of-line (HOL) blocking. This is partly a fundamental limitation of
+    TCP, i.e. there's no virtual streams. Such limitations are overcome in QUIC and HTTP/3.)
+    Either of these 2 approaches may work:
+    - Only read socket when we finish processing a message. Pub-Sub app may queue up the frame for later processing, or it may immedieiately process it.
+    - Give FS2-Netty a queue to fill and only backpressure when queue is full (or above a high watermark, start reading when below a low watermark).
+- We want to send ws frames. 
+  - When sending any frame, we want to know if it was written to the connection. 
+  - When sending a close frame, then the frame itself should be sent, then the connection should be closed.
+  - We want to know if we're writing too fast and respond to backpressure (namely TCP backpressure).
+- We want to know why a websocket connection closed, what was the close code if any.
+
+WebSocket (extras that will save us time dev time if framework can do these):
+- We want to group connections, label an individual connection, write to a single connection, broadcast to group, close a single connection, and close all connections.
+- We want to know if connection is still alive.
+  - Close connection if it's unresponsive.
+- We want metrics for WS server
+  - Count of different frame types
+  - The size of data frames
+  - Count of different close codes
+    
+General Server configs:
+- We want to limit the number of TCP connections? To protect the server?
+    - Backpressure accepting connections (and reads) if memory is getting high?
+      E.g. Netty thread pool is queuing too many tasks in compute thread pool. Queue will hit OOM.
+
+## Implementation thoughts for above use cases
+Based on what currently exists and past discussions with additional modifications.
+
+## Package Structure
+- fs2-netty-core
+  - Netty
+    - wrappers
+    - ops
+- fs2-netty-server
+  - tcp
+  - udp
+- fs2-netty-client
+  - tcp
+  - udp
+- fs2-netty-all
+- fs2-netty-simple-http-server
+  - websockers
```

Comment:
I like the idea of having some nicer wrappers around the existing pipeline handlers in these areas. So exactly what you're saying here, I think. :-) We should still allow users to wrap their own arbitrary handlers if they need to, but nice experiences out of the box are important.

Reply:
yeah exactly, there's the core and some user-friendly default pipelines to make people's lives easier.